crates/uv-auth/src/middleware.rs

```diff
@@ -100,88 +140,199 @@ impl Middleware for AuthMiddleware {
+
+        // Otherise, attempt an anonymous request
+        trace!("Attempting unauthenticated request for {url}");
+
+        // <https://github.com/TrueLayer/reqwest-middleware/blob/abdf1844c37092d323683c2396b7eefda1418d3c/reqwest-retry/src/middleware.rs#L141-L149>
+        // Clone the request so we can retry it on authentication failure
+        let mut retry_request = request.try_clone().ok_or_else(|| {
+            Error::Middleware(anyhow!(
+                "Request object is not clonable. Are you passing a streaming body?".to_string()
+            ))
+        })?;
+
+        let response = next.clone().run(request, extensions).await?;
+
+        // And on failure, search for credentials
+        if matches!(
+            response.status(),
+            StatusCode::FORBIDDEN | StatusCode::NOT_FOUND | StatusCode::UNAUTHORIZED
+        ) {
+            trace!(
+                "Request for {url} failed with {}, checking for credentials",
+                response.status()
+            );
+            let credentials = self.cache().get_realm(
+                Realm::from(retry_request.url()),
+                credentials
+                    .map(|credentials| credentials.to_username())
+                    .unwrap_or(Username::none()),
+            );
+
+            if let Some(credentials) = credentials.as_ref() {
+                if credentials.password().is_some() {
+                    trace!(
+                        "Retrying request for {url} with credentials from cache {credentials:?}"
+                    );
+                    retry_request = credentials.authenticate(retry_request);
+                    return self
+                        .complete_request(None, retry_request, extensions, next)
+                        .await;
+                }
+            }
+
+            if let Some(credentials) = self
+                .fetch_credentials(credentials.as_deref(), retry_request.url())
+                .await
+            {
+                retry_request = credentials.authenticate(retry_request);
+                trace!("Retrying request for {url} with {credentials:?}");
+                return self
+                    .complete_request(Some(Arc::new(credentials)), retry_request, extensions, next)
+                    .await;
+            } else {
+                if let Some(credentials) = credentials.as_ref() {
+                    if !attempt_has_username {
+                        trace!(
+                            "Retrying request for {url} with username from cache {credentials:?}"
+                        );
+                        retry_request = credentials.authenticate(retry_request);
+                        return self
+                            .complete_request(None, retry_request, extensions, next)
+                            .await;
+                    }
+                }
+
+                Ok(response)
+            }
+        } else {
+            Ok(response)
```

Comment:
Maybe more of a personal preference thing, but I find long if blocks trailed by a short/trivial else block to be more difficult to read. I'm more of a fan of using early returns to simplify (which are used above here). So in this case, maybe:

```rust
if !matches!(
    response.status(),
    StatusCode::FORBIDDEN | StatusCode::NOT_FOUND | StatusCode::UNAUTHORIZED
) {
    return Ok(response);
}

trace!(
    "Request for {url} failed with {}, checking for credentials",
    response.status()
);
// and so on...
```

Basically the idea is that when you're reading the code in a linear fashion, you can check the condition, note the early return and then move on. But with the code written this way, you kinda have to keep the condition in mind as you go on, and then we you get to the `else`, I find I usually have to go back up to remind myself of what the condition is.

The other upshot of this is that you get less nesting.

Reply:
THanks!